File: library/src/main/java/com/rejasupotaro/octodroid/GitHub.java
package com.rejasupotaro.octodroid;

import com.google.gson.Gson;
import com.rejasupotaro.octodroid.http.ApiClient;

public class GitHub {
    private static GitHubClient client;

    public static void initialize(ApiClient apiClient) {
        if (client == null) {
            synchronized (GitHub.class) {
                if (client == null) {
                    client = new GitHubClient(apiClient);
                }
            }
        }
    }

    public static GitHubClient client() {
        if (client == null) {
            synchronized (GitHub.class) {
                if (client == null) {
                    client = new GitHubClient(new ApiClient());
                }
            }
        }
        return client;
    }

    public static void gson(Gson gson) {
        GsonProvider.set(gson);
    }
}

File: app/src/main/java/com/example/octodroid/data/GitHub.kt
package com.example.octodroid.data

import com.google.gson.Gson
import com.rejasupotaro.octodroid.GitHubClient
import com.rejasupotaro.octodroid.GsonProvider
import com.rejasupotaro.octodroid.http.ApiClient

object GitHub {
    private var client: GitHubClient? = null

    fun initialize(apiClient: ApiClient) {
        if (client == null) {
            synchronized (GitHub::class.java) {
                if (client == null) {
                    client = GitHubClient(apiClient)
                }
            }
        }
    }

    fun client(): GitHubClient {
        if (client == null) {
            synchronized (GitHub::class.java) {
                if (client == null) {
                    client = GitHubClient(ApiClient())
                }
            }
        }
        return client
    }

    fun gson(gson: Gson) {
        GsonProvider.set(gson)
    }
}
